src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -94,6 +101,7 @@ internal AssemblyLoadContext(bool fRepresentsTPALoadContext, bool isCollectible)
         ~AssemblyLoadContext()
         {
             // Only valid for a Collectible ALC. Non-collectible ALCs have the finalizer suppressed.
+            Debug.Assert(IsCollectible);
```

Comment:
Note that you can get here for completely uninitialized objects, e.g. when the JITing of a constructor for a custom ALC fails: We have allocated the object by the time we are JITing the constructor already, but none of the constructor has run.

Reply:
Hmm, I didn't know that. I assume then that this could possibly be the real reason behind #21100. Then it seems that I could use e.g. the `unloadLock` as a guard (if it is null, the constructor was not executed and thus we should skip the whole finalizer code)